File: phi_plugin/model/cls/LevelRecordInfo.ts
import fCompute from '../fCompute';
import getInfo from '../getInfo';
import { idString } from '../type/type';

export default interface LevelRecordInfo {
    id: idString
    fc: boolean
    score: number
    acc: number
    rank: string
    song: string
    illustration: string
    Rating: string
    difficulty: number
    rks: number
    suggest?: string
}

export default class LevelRecordInfo {


    /**
     * @param data 原始数据
     * @param id 曲目id
     * @param rank 难度
     */
    constructor(data: { fc: boolean; score: number; acc: number; }, id: idString, rank: number) {


        this.fc = data.fc;
        this.score = data.score;
        this.acc = data.acc;
        this.id = id as idString

        let info = getInfo.info(id)

        if (!info) {
            return
        }

        this.rank = getInfo.Level[rank] //AT IN HD EZ LEGACY 
        this.song = info.song //曲名
        this.illustration = getInfo.getill(id) //曲绘链接
        this.Rating = Rating(this.score, this.fc) //V S A 


        if (info.chart && info.chart[this.rank]?.difficulty) {
            this.difficulty = Number(info.chart[this.rank]['difficulty']) //难度
            this.rks = fCompute.rks(this.acc, this.difficulty) //等效rks
        } else {
            this.difficulty = 0
            this.rks = 0
        }


    }
}

function Rating(score, fc) {
    if (score >= 1000000)
        return 'phi'
    else if (fc)
        return 'FC'
    else if (!score)
        return 'NEW'
    else if (score < 700000)
        return 'F'
    else if (score < 820000)
        return 'C'
    else if (score < 880000)
        return 'B'
    else if (score < 920000)
        return 'A'
    else if (score < 960000)
        return 'S'
    else
        return 'V'
}

File: phi_plugin/model/LevelRecordInfo.py
from ..utils import Rating
from .cls.LevelRecordInfo import LevelRecordInfoModel
from .cls.type import idString
from .fCompute import fCompute
from .getInfo import getInfo


class LevelRecordInfo:
    async def constructor(
        self, params: dict, id: idString, rank: int
    ) -> LevelRecordInfoModel | None:
        """
        :param params: 原始数据
        :param id: 曲目id
        :param rank: 难度
        """
        data = {
            "fc": params["fc"],
            "score": params["score"],
            "acc": params["acc"],
            "id": id,
        }
        info = await getInfo.info(id)
        if not info:
            return None
        data["rank"] = getInfo.Level[rank]  # AT IN HD EZ LEGACY
        data["song"] = info.song  # 曲名
        data["illustration"] = await getInfo.getill(id)  # 曲绘链接
        data["Rating"] = Rating(data["score"], data["fc"])  # V S A
        if (
            info.chart
            and info.chart[data["rank"]]
            and info.chart[data["rank"]].difficulty
        ):
            data["difficulty"] = info.chart[data["rank"]].difficulty  # 难度
            data["rks"] = fCompute.rks(data["acc"], data["fc"])  # 等效rks
        else:
            data["difficulty"] = 0
            data["rks"] = 0
        return LevelRecordInfoModel(**data)
